### ISSUE TITLE
More flexibile way to specify MaxTime

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#2616]
Currently, we can use `MaxTimeAttribute` to specify the length of time that a given test is supposed to take. If the test succeeds but takes more than the specified maximum, we fail it with a special message.

This works well form any things and should be retained. However, we need flexibility on two dimensions:

1. Granularity, that is being able to specify a maximum time for a specific bit of code within hte test rather than just the whole test.

2. Ability to specify Warning as the outcome rather than failure.

I think that we need to make some decisions on the first item before we can deal with the second. That requires a decision: Do we want to be able to specify a maximum time for a single assertion only or do we also want to be able to do it for arbitrary code?

If we want to do it for arbitrary code, then we need to consider different mechanisms from what would be required to do it for a constraint or an assertion.

Thoughts?

To get things going, consider the following as just some possible options...

```C#
Assert.That(<somecode>, Has.MaxTime(ms));
Assert.That(<somecode>, Has.Duration.LessThan(ms));
Assert.That(<somecode>, <someassertion>.MaxTime(ms));
Assert.Maxtime(<somecode>, ms);
```

**Comment by jnm2** [comment]
The solution space will probably overlap, but can you add the dimension of able to change the timeout amount at runtime?

**Comment by CharliePoole** [comment]
I figured it was implied, but yes, that is another dimension. For example, in the samples I threw out, the value of `ms` could either be constant or vary at runtime.

**Comment by jnm2** [comment]
If we added `Assert.MaxTime`, what would we do to accommodate warning?
I like `Has.Duration.LessThan` because it's more composable if you're actually expecting `MoreThan`. `MaxTime` is more recognizable.

The assertion or constraint must be applied to a delegate. I could foresee composing `MaxTime` both ways:
```cs
Assert.That(del, Has.MaxTime(1000).And.ReturnValue.EqualTo("foo"));
Assert.That(obj, Has.Property("FooDelegate").With.MaxTime(1000));
```

**Comment by jnm2** [comment]
@CharliePoole It seems you and I are the only team members that have this issue on our radar.

> What do we want to enable users to measure? What kinds of measurement are within our product vision and what kinds are not?

I'd say we want to enable measuring the time it takes an arbitrary synchronous or asynchronous action to execute, as precisely as possible. Examples might be database, calculation or user interface responsiveness. You might also want to make sure something doesn't happen before a certain time. You might want to use a tolerance.

**Comment by jnm2** [comment]
This use case crossed my mind, and then I just saw that someone asked for it! https://github.com/nunit/nunit/issues/1460

Do we want to add the ability to measure the time it takes for a given event or change to occur? This is very similar to waiting for asynchronous code, but we'd probably provide a nice API so that people don't have to write TaskCompletionSource boilerplate.

**Comment by CharliePoole** [comment]
@jnm2 I agree with the idea of being able to measure arbitrary code and the need to test duration against various criteria like more than or less than. I hadn't thought about synchronous versus asynchronous but yes... it would be nice to allow the same operations on both. If the user supplied a synchronous delegate, we could add a layer of async on top of it, saving them the trouble.

OTOH, I don't think we should mix time measurement with timeout on an assert statement. I think we should keep timeout at the test method level. If the user has other long operations in the same test then they should either (1) NOT do that! or (2) count the time appropriately. Obviously, it's __possible__ to move timeout to the statement level, but I don't see enough benefit to justify the effort and the complication it would bring to the code if any assert could time out.

Turning the page to coding...

Implementation-wise, I'd see both a Duration constraint and an Assert, analogous to how Throws works. The constraint, like Throws would be non-chainable, making it easier to understand and requiring an extra step to test both the result and the duration. Examples...

```C#
Assert.Duration(del, LessThan(1000));
Assert.That(del, Has.Duration.LessThan(1000);
var duration = Assert.Duration(del);
Assert.That(duration, Is.LessThan(1000);

Assert.That(() => Assert.That (del, Is.EqualTo("foo")), Has.Duration.LessThan(1000));
var duration = Assert.Duration( Assert.That(del, Is.EqualToFoo));
Assert.That(duration, Is.LessThan(1000));

Assert.That(() => Assert.That(del, Throws.InstanceOf<FooException>()), Has.Duration.MoreThan(1000));
var duration = Assert.Duration(() => Assert.Throws(del, InstanceOf<FooException>()));
Assert.That(duration, Is.GreaterThan(1000));

Assert.That(() => Assert.That(obj, Has.Property("FooDelegate")), Has.Duration.LessThan(1000));
var duration = Assert.Duration(() => Assert.That(obj, Has.Property("FooDelegate")));
Assert.That(duration, Is.LessThan(1000));
```

As with Throws, I think many people will prefer the clarity of two separate statements.

**Comment by jnm2** [comment]
The only scenario that I can think of where method-level timeout granularity is insufficient is if I need to timeout with a `try`...`finally` and guarantee that cleanup happens. Knowing NUnit's internal timeout mechanism, I know that `try`...`finally` will still work, but semantically there is no guarantee. The framework might switch to some other abortion tactic in the future and the clean up would stop happening as a side effect of the ThreadAbortException. This is the only thing I can think of that gets the semantics _right_:

```cs
[Test]
public void Foo()
{
    // Set static/machine global state
    try
    {
        Assert.Timeout(() => SomeUnreliableUIOperation(42), TimeSpan.FromSeconds(30));
    }
    finally
    {
        // restore static/machine global state
    }
}
```

That said, we could brush this off till it actually causes real-world pain.

-----

I like all of this:

```cs
Assert.Duration(() => foo.Bar(42), LessThan(1000));
Assert.That(() => foo.Bar(42), Has.Duration.LessThan(1000);
var duration = Assert.Duration(() =>foo.Bar(42));
Assert.That(duration, Is.LessThan(1000);
```

The assert nested in another assert's lambda is not my favorite. The syntax makes it harder for me to think about what's happening:
```cs
Assert.That(() => Assert.That(() => foo.Bar(42), Is.EqualTo("foo")), Has.Duration.LessThan(1000));
var duration = Assert.Duration(Assert.That(() => foo.Bar(42), Is.EqualToFoo));
Assert.That(duration, Is.LessThan(1000));
```

But since we get that nesting possibility automatically once we add the first syntax, it's a good incremental first step which covers all bases. If we find ourselves wanting to combine duration asserts with other asserts we can revisit.

**Comment by CharliePoole** [comment]
I've used nested asserts in a lot of our tests and it's not so bad when you use good spacing and indentation, something I didn't bother with in my examples. For example...

```C#
Assert.That(() => 
    Assert.That(() => foo.Bar(42), Is.EqualTo("foo")), 
    Has.Duration.LessThan(1000));

var duration = Assert.Duration(Assert.That(() => foo.Bar(42), Is.EqualToFoo));
Assert.That(duration, Is.LessThan(1000));
```

These don't read quite so bad.

Another option we discussed a long time ago is to have a `MaxTime` constraint modifier, working syntactically like the existing `After` modifier.

```C#
Assert.That(() => foo.Bar(42), Is.EqualTo("foo").MaxTime(1000));
```C#

This is simpler and easier to implement but limits us to specifying a maximum. It's even possible to create a modifier that works like this:

```C#
Assert.That(() => foo.Bar(42), Is.EqualTo("foo").Duration.LessThanOrEqualTo(1000));
```

However the latter is a lot more work and would be the first instance of a modifier that took a constraint. In fact, it might be better to abandon the Modifier syntax and make Duration a constraint in that case.

The trouble with both these examples, however, is that they can __only__ work with assertions, not with general code instances.

My inclination is to start out with `Assert.Duration`, returning the duration value over an arbitrary delegate. Then we could add the `Has.Duration` constraint. Some sort of shorthand modifier could wait till we see a real need. I've learned that adding features too quickly, one on top of the other, often leads us to later make breaking changes.

**Comment by jnm2** [comment]
Okay. I agree with that. Are we sure enough of this direction that I should go ahead and close https://github.com/nunit/nunit/pull/2513?

**Comment by CharliePoole** [comment]
@jnm2 I don't want to be the one to answer that, given my lame-duckiness. 😄

**Comment by jnm2** [comment]
@rprouse What do you think?

**Comment by jnm2** [comment]
@CharliePoole Here's a feature request for a timeout that's more granular than we allow today: https://github.com/nunit/nunit/issues/2397

What if we added `Assert.Timeout` rather than the `IncludeInSetupAndTearDown` flag to address it?

**Comment by CharliePoole** [comment]
@jnm2 That would solve the need even though it's not the feature requested. However, let's not mix it up with MaxTime. Timeout is a lot more complicated.
